File: app/src/main/java/com/mlsdev/mapsappsample/placesuggestions/PlacesAdapter.java
package com.mlsdev.mapsappsample.placesuggestions;

import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mlsdev.mapsappsample.R;
import com.mlsdev.mapsappsample.databinding.PlaceSuggestionsItemBinding;

import java.util.ArrayList;
import java.util.List;

public class PlacesAdapter extends RecyclerView.Adapter<PlacesAdapter.ViewHolder> {
    private List<String> places = new ArrayList<>();
    private OnItemClickListener onItemClickListener;

    public PlacesAdapter(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    @Override
    public PlacesAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        PlaceSuggestionsItemBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()),
                R.layout.place_suggestions_item,
                parent,
                false
        );

        PlaceViewModel viewModel = new PlaceViewModel(onItemClickListener);
        binding.setViewModel(viewModel);
        return new ViewHolder(binding.getRoot(), viewModel);
    }

    @Override
    public void onBindViewHolder(PlacesAdapter.ViewHolder holder, int position) {
        holder.viewModel.placeFullText.set(places.get(position));
    }

    @Override
    public int getItemCount() {
        return places.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private PlaceViewModel viewModel;

        public ViewHolder(View itemView, PlaceViewModel viewModel) {
            super(itemView);
            this.viewModel = viewModel;
        }
    }

    public void setData(List<String> places) {
        this.places = places;
        notifyDataSetChanged();
    }

    public void clearData() {
        places.clear();
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/mlsdev/mapsappsample/placesuggestions/PlacesAdapter.kt
package com.mlsdev.mapsappsample.placesuggestions

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.mlsdev.mapsappsample.databinding.PlaceSuggestionsItemBinding
import java.util.*

class PlacesAdapter(
        private val onItemClickListener: (place:String) -> Unit
) : RecyclerView.Adapter<PlacesAdapter.ViewHolder>() {
    private var places: MutableList<String> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = PlaceSuggestionsItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)

        val viewModel = PlaceViewModel(onItemClickListener)
        binding.viewModel = viewModel
        return ViewHolder(binding.root, viewModel)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.viewModel.placeFullText.set(places[position])
    }

    override fun getItemCount(): Int {
        return places.size
    }

    inner class ViewHolder(itemView: View, val viewModel: PlaceViewModel) : RecyclerView.ViewHolder(itemView)

    fun setData(places: MutableList<String>) {
        this.places = places
        notifyDataSetChanged()
    }

    fun clearData() {
        places.clear()
        notifyDataSetChanged()
    }
}
